app/grandchallenge/discussion_forums/views.py

```diff
@@ -108,3 +118,116 @@ def get_success_url(self):
             "discussion-forums:topic-list",
             kwargs={"challenge_short_name": self.request.challenge.short_name},
         )
+
+
+class ForumPostCreate(ObjectPermissionRequiredMixin, CreateView):
+    model = ForumPost
+    permission_required = "discussion_forums.create_topic_post"
+    raise_exception = True
+    form_class = ForumPostForm
+
+    @cached_property
+    def forum(self):
+        return self.request.challenge.discussion_forum
+
+    @cached_property
+    def topic(self):
+        return get_object_or_404(
+            ForumTopic, forum=self.forum, slug=self.kwargs["slug"]
+        )
+
+    def get_permission_object(self):
+        return self.topic
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data()
+        context.update({"forum": self.forum, "topic": self.topic})
+        return context
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs.update({"user": self.request.user, "topic": self.topic})
+        return kwargs
+
+
+class ForumPostDetail(ObjectPermissionRequiredMixin, DetailView):
+    model = ForumPost
+    permission_required = "discussion_forums.view_forumpost"
+    raise_exception = True
+
+    @cached_property
+    def forum(self):
+        return self.request.challenge.discussion_forum
+
+    def get_object(self, queryset=None):
+        return get_object_or_404(
+            ForumPost,
+            pk=self.kwargs["pk"],
```

Comment:
It may not be an issue but for these three views only filtering by pk and not forum will allow the views in the context of another challenge or topic. You may want to add the `topic__forum=self.forum` filter too.